File: app/schemas/menu.py
"""
菜单Schema
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class MenuBase(BaseModel):
    """菜单基础Schema"""
    name: str = Field(..., min_length=1, max_length=100, description="菜单名称")
    path: Optional[str] = Field(None, max_length=200, description="路由路径")
    icon: Optional[str] = Field(None, max_length=50, description="图标名称（Ant Design图标名称）")
    permission_code: Optional[str] = Field(None, max_length=100, description="关联的权限码（type='menu'的权限）")
    sort_order: int = Field(0, description="排序序号")
    visible: bool = Field(True, description="是否可见")
    parent_id: Optional[str] = Field(None, description="父菜单ID")


class MenuCreate(MenuBase):
    """创建菜单Schema"""
    tenant_id: Optional[str] = Field(None, description="租户ID（None表示系统级菜单）")


class MenuUpdate(BaseModel):
    """更新菜单Schema"""
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="菜单名称")
    path: Optional[str] = Field(None, max_length=200, description="路由路径")
    icon: Optional[str] = Field(None, max_length=50, description="图标名称")
    permission_code: Optional[str] = Field(None, max_length=100, description="关联的权限码")
    sort_order: Optional[int] = Field(None, description="排序序号")
    visible: Optional[bool] = Field(None, description="是否可见")
    parent_id: Optional[str] = Field(None, description="父菜单ID")
    status: Optional[bool] = Field(None, description="状态：True-启用，False-停用")


class MenuResponse(MenuBase):
    """菜单响应Schema"""
    id: str
    tenant_id: Optional[str]
    status: bool
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str]
    
    class Config:
        from_attributes = True


class MenuTreeItem(MenuResponse):
    """菜单树节点Schema"""
    children: Optional[List['MenuTreeItem']] = None


class MenuTreeResponse(BaseModel):
    """菜单树响应Schema"""
    items: List[MenuTreeItem]


class MenuSortRequest(BaseModel):
    """菜单排序请求Schema"""
    menu_orders: List[dict] = Field(..., description="菜单排序列表，格式：[{'id': 'menu_id', 'sort_order': 0}]")


# 更新前向引用
MenuTreeItem.model_rebuild()

File: frontend/src/api/menu.ts
import client from './client'

export interface MenuItem {
  id: string
  parent_id?: string
  name: string
  path?: string
  icon?: string
  permission_code?: string
  sort_order: number
  visible: boolean
  tenant_id?: string
  status: boolean
  created_at: string
  updated_at: string
  created_by?: string
  children?: MenuItem[]
}

export const menuApi = {
  // 获取当前用户的可见菜单（动态菜单接口）
  getMyMenus: async () => {
    const response = await client.get<MenuItem[]>('/menus/my')
    return response.data
  },

  // 获取菜单列表（树形结构）
  getMenus: async (tenantId?: string, includeInvisible = false) => {
    const params = new URLSearchParams()
    if (tenantId) {
      params.append('tenant_id', tenantId)
    }
    if (includeInvisible) {
      params.append('include_invisible', 'true')
    }
    const queryString = params.toString()
    const url = queryString ? `/menus?${queryString}` : '/menus'
    const response = await client.get<MenuItem[]>(url)
    return response.data
  },

  // 获取单个菜单
  getMenu: async (menuId: string) => {
    const response = await client.get<MenuItem>(`/menus/${menuId}`)
    return response.data
  },

  // 创建菜单
  createMenu: async (menu: Partial<MenuItem>) => {
    const response = await client.post<MenuItem>('/menus', menu)
    return response.data
  },

  // 更新菜单
  updateMenu: async (menuId: string, menu: Partial<MenuItem>) => {
    const response = await client.put<MenuItem>(`/menus/${menuId}`, menu)
    return response.data
  },

  // 删除菜单
  deleteMenu: async (menuId: string) => {
    await client.delete(`/menus/${menuId}`)
  },

  // 启用/停用菜单
  updateMenuStatus: async (menuId: string, status: boolean) => {
    const response = await client.patch<MenuItem>(`/menus/${menuId}/status?status=${status}`)
    return response.data
  },

  // 批量更新菜单排序
  updateMenuSort: async (menuOrders: Array<{ id: string; sort_order: number }>) => {
    await client.post('/menus/sort', { menu_orders: menuOrders })
  },
}

